logstash-core/lib/logstash/pipeline.rb

```diff
@@ -222,73 +228,43 @@ def start_workers
-    namespace_pipeline = metric.namespace([:stats, :pipelines, pipeline_id.to_s.to_sym, :events])
+    @filter_queue_client.set_batch_details(batch_size, batch_delay)
 
     while running
-      # To understand the purpose behind this synchronize please read the body of take_batch
-      input_batch, signal = @input_queue_pop_mutex.synchronize { take_batch(batch_size, batch_delay) }
+      batch, signal = @filter_queue_client.take_batch
+      @events_consumed.increment(batch.size)
       running = false if signal == LogStash::SHUTDOWN
-
-      @events_consumed.increment(input_batch.size)
-      namespace_events.increment(:in, input_batch.size)
-      namespace_pipeline.increment(:in, input_batch.size)
-
-      filtered_batch = filter_batch(input_batch)
+      filter_batch(batch)
 
       if signal # Flush on SHUTDOWN or FLUSH
         flush_options = (signal == LogStash::SHUTDOWN) ? {:final => true} : {}
-        flush_filters_to_batch(filtered_batch, flush_options)
+        flush_filters_to_batch(batch, flush_options)
       end
 
-      @events_filtered.increment(filtered_batch.size)
-
-      namespace_events.increment(:filtered, filtered_batch.size)
-      namespace_pipeline.increment(:filtered, filtered_batch.size)
-
-      output_batch(filtered_batch)
+      output_batch(batch)
 
-      namespace_events.increment(:out, filtered_batch.size)
-      namespace_pipeline.increment(:out, filtered_batch.size)
-
-      inflight_batches_synchronize { set_current_thread_inflight_batch(nil) }
-    end
-  end
+      # @filter_queue_client.ack(batch)
+      # @dlq_client.add(batch)
 
-  def take_batch(batch_size, batch_delay)
-    batch = []
-    # Since this is externally synchronized in `worker_look` wec can guarantee that the visibility of an insight batch
-    # guaranteed to be a full batch not a partial batch
-    set_current_thread_inflight_batch(batch)
-
-    signal = false
-    batch_size.times do |t|
-      event = (t == 0) ? @input_queue.take : @input_queue.poll(batch_delay)
-
-      if event.nil?
-        next
-      elsif event == LogStash::SHUTDOWN || event == LogStash::FLUSH
-        # We MUST break here. If a batch consumes two SHUTDOWN events
-        # then another worker may have its SHUTDOWN 'stolen', thus blocking
-        # the pipeline. We should stop doing work after flush as well.
-        signal = event
-        break
-      else
-        batch << event
-      end
+      @filter_queue_client.close_batch(batch)
     end
-
-    [batch, signal]
   end
 
   def filter_batch(batch)
-    batch.reduce([]) do |acc,e|
-      if e.is_a?(LogStash::Event)
-        filtered = filter_func(e)
-        filtered.each {|fe| acc << fe unless fe.cancelled?}
+    batch.each do |event|
+      if event.is_a?(LogStash::Event)
+        filter_func(event).each do |e|
+          #these are both dequeued and generated events
+          if e.cancelled?
+            batch.cancel(e)
+          else
+            batch.add(e)
```

Comment:
also `#dequeue(event)` is really confusing  because it does not do what the method name generally means: to remove an element from a queue, but adds the event into a `dequeued` hash. 

I maintain that `add` and `dequeue` are very confusing method names - it may seems clear for you now but forward looking for someone reading the pipeline code in the future willl have a very hard time understanding this because we reuse standard method names with very typical meaning and do something completely counter intuitive.